### PR TITLE
Thread-safe Hash implementation: fix for issue/565

### PR DESCRIPTION
**Pull request by gjeee**
I changed the Hash implementation. I made sure that all the constructors read all the state vars in 1 go in a synchronised fashion. It fixes issue 565 https://github.com/iotaledger/iri/issues/565

### CODE REVIEW

File: src/main/java/com/iota/iri/model/Hash.java [hunk 1]
```diff
@@ -19,33 +21,56 @@
     private byte[] bytes;
     private int[] trits;
```
Comment (gubatron) [inline src/main/java/com/iota/iri/model/Hash.java]: why not just make these 2 guys final, and use new instances whenever you need a Hash, nothing to put synchronized sections on, just make a new copy.

In my experience, synchronized should be a last resource. Try to work it functional, no mutability.
All those synchronized blocks most likely will end up creating deadlock headaches.

File: src/main/java/com/iota/iri/model/Hash.java [hunk 2]
```diff
@@ -19,33 +21,56 @@
...
     private int hashCode;
+
```
Comment (gubatron) [inline src/main/java/com/iota/iri/model/Hash.java]: have you tried making these 3 properties final, calculated when the object is created.
The best way to make code thread safe is to make it immutable. If you need to change it in any way, just create a new instance, the functional approach is way more robust and easier to parallelize.

In my experience, `synchronized` should be a last resource and usually a sign of something bad in design. 
Try to work it functionally, with no mutability.
All those synchronized blocks most likely will end up creating deadlock headaches.

Comment (gjeee) [reply src/main/java/com/iota/iri/model/Hash.java]: hi tx for your suggestion.....

like u said maybe setting all properties in the constructor and make them final can make this whole class thread-safe.....of course there is no lazy initialization but at least coding-wise a lot simpler....

but there is already another PR as a possible fix https://github.com/iotaledger/iri/pull/577
..... it features lazy initialization but it still needs some sync blocks though....so all in all a bit more complex

Comment (gubatron) [reply src/main/java/com/iota/iri/model/Hash.java]: wish I had more time to try and code this myself immutably, perhaps a few cascading refactors might show weaknesses in the architecture

**Comment by GalRogozinski** [comment]
Close this PR in favor of #577

**Pull request closed without merging**